### ISSUE TITLE
PKCS11#C_WaitForSlotEvent cannot detect smart card removal for Yubikey device on Ubuntu 18.04

### ISSUE THREAD

**Issue by Sabrina-Hou** [OpenSC/OpenSC#1942]
### Proposed Resolution
For Yubikey smart card, it may return SCARD_E_UNKNOWN_READER error when it invokes SCardGetStatusChange function when the smart card is removed. For the function refresh_attributes, it needs to handle the error code SCARD_E_UNKNOWN_READER after invoking SCardGetStatus and set the reader->flags |= SC_READER_CARD_CHANGED.
https://github.com/OpenSC/OpenSC/blob/0fd77d642c1e7f3cff6f9cf4b60da0d5e516bbe4/src/libopensc/reader-pcsc.c#L366

Because the function refresh_attributes cannot set the reader's flag correctly, this raises the C_WaitForSlotEvent function cannot detect smart card removal status.
https://github.com/OpenSC/OpenSC/blob/bfa8415ea3f5a0b055ef383863e2711b95b04ec8/src/pkcs11/pkcs11-global.c#L712

### Steps to reproduce
1. Plugin Yubikey smart card to physical Ubuntu 18.04 machine
2. Invoke C_WaitForSlotEvent to regularly check smart card removal status
3. Unplug Yubikey from machine
Expected: The function C_WaitForSlotEvent can return smart card removal state
Actual result: The function C_WaitForSlotEvent return CKR_NO_EVENT state.

### Logs
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] pkcs11-global.c:653:C_WaitForSlotEvent: C_WaitForSlotEvent(block=0)
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] slot.c:518:slot_find_changed: called
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] slot.c:389:card_detect_all: Detect all cards
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] slot.c:235:card_detect: Yubico Yubikey 4 OTP+U2F+CCID 00 00: Detecting smart card
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] sc.c:275:sc_detect_card_presence: called
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] reader-pcsc.c:411:pcsc_detect_card_presence: called
0x7f0810575b00 11:52:29.373 [opensc-pkcs11] reader-pcsc.c:319:refresh_attributes: Yubico Yubikey 4 OTP+U2F+CCID 00 00 check
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] reader-pcsc.c:341:refresh_attributes: Yubico Yubikey 4 OTP+U2F+CCID 00 00:SCardGetStatusChange failed: 0x80100009
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] reader-pcsc.c:415:pcsc_detect_card_presence: returning with: -1900 (Unknown error)
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] sc.c:280:sc_detect_card_presence: returning with: -1900 (Unknown error)
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:240:card_detect: Yubico Yubikey 4 OTP+U2F+CCID 00 00: failed, Unknown error
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: -1900 (Unknown error)
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:408:card_detect_all: All cards detected
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:525:slot_find_changed: slot 0x0 token: 1 events: 0x00
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:531:slot_find_changed: mask: 0x0F events: 0x00 result: 0
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:525:slot_find_changed: slot 0x1 token: 0 events: 0x00
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:531:slot_find_changed: mask: 0x0F events: 0x00 result: 0
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:525:slot_find_changed: slot 0x2 token: 0 events: 0x00
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:531:slot_find_changed: mask: 0x0F events: 0x00 result: 0
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:525:slot_find_changed: slot 0x3 token: 0 events: 0x00
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:531:slot_find_changed: mask: 0x0F events: 0x00 result: 0
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] slot.c:539:slot_find_changed: returning with: 8
0x7f0810575b00 11:52:29.374 [opensc-pkcs11] pkcs11-global.c:710:C_WaitForSlotEvent: C_WaitForSlotEvent() = CKR_NO_EVENT

**Comment by frankmorgner** [comment]
I think I've experienced this earlier while preparing https://github.com/OpenSC/OpenSC/pull/1923. The problem is that SCARD_E_UNKNOWN_READER is thrown if one reader (of many available readers) in the list disappears. The solution in #1923 was to re-initialize the list of readers (sc_detect_readers) and to check which one disappeared. Note that #1923 already changes the reader's flag to set SC_READER_REMOVED correctly. I think I've catched all corner cases, but unfortunately I didn't find the time to test this on Windows.

I originally thought about rewriting sc_wait_for_event to handle this complexity at a lower level, but didn't want to touch PKCS#11. Since you are reporting the same error here, too, I think we should give it a second thought...
